Design/Instrument-2.0/design-notes.md

```diff
@@ -0,0 +1,89 @@
+
+## Overview
+
+- Copy-on-write `InstrumentTree`, similar to the [prototype](https://github.com/DMSC-Instrument-Data/instrument-prototype). The design of this component is not the core focus of this document.
+- `DetectorInfo`, a class that stores (and provides access to) basic parameters of detectors, such as L2, twoTheta, position, masking flags, monitor flags, ....
+- `SpectrumInfo`, a class that stores (and provides access to) basic parameters of spectra, such as L2, twoTheta, position, masking flags, monitor flags, .... The values hare are derived from those on `DetectorInfo`. This class reflects the common case in Mantid that a spectrum can represent more than one detector, e.g., due to grouping of detectors or after summing spectra.
+
+
+## Notes
+
+### Spectrum-number and detector ID
+
+Mantid currently works with `SpectrumNumber` and `DetectorID` in many places.
+As discussed in detail in this [design proposal](https://github.com/mantidproject/documents/blob/spectrum_number_and_workspace_index_abstraction/Design/spectrum_number_and_workspace_index_abstraction.md), these concepts should not be spread accross our codebase.
+Apart from the arguments presented there, early tests with the [prototype](https://github.com/DMSC-Instrument-Data/instrument-prototype) showed that using `DetectorID` as an index for accessing detector data is inefficient, since the resulting `std::map` lookups slow things down significantly.
+
+Therefore, neither `SpectrumNumber` nor `DetectorID` should be part of the core mechanics of the instrument.
+Translation to a plain index, e.g., for accessing a `std::vector` should happen as close to the user interface as possible.
+Mechanism for this translation would probably be part of the workspace interface, but separate from, e.g., `SpectrumInfo`.
+The remainder of this document is thus not concerned with `SpectrumNumber` or `DetectorID`.
+
+### Naming
+
+All class names in this document are preliminary and just for the sake of discussion.
+Better names are needed in several cases.
+
+
+## Design
+
+### Structure
+
+Algorithms that work with workspaces need to access information about spectra.
+Therefore:
+
+- `SpectrumInfo` should be a member of workspace.
+- `SpectrumInfo` owns `DetectorInfo`.
+- `DetectorInfo` owns `InstrumentTree`.
+
+If no detectors are grouped, there is a 1:1 mapping between spectrum and detector.
+`SpectrumInfo` and `DetectorInfo` should thus share data, to avoid waste of memory and CPU resources.
+This can be facilitated by:
+
+- `GeometryDataArray` is used to store vectors of data, such as L2, twoTheta, position, masking flags, monitor flags, ....
+- `SpectrumInfo` and `DetectorInfo` have a field `std::shared_ptr<GeometryDataArray>`. By default both point to the same instance. If detectors are grouped, `SpectrumInfo` allocates its own `GeometryDataArray`.
+
+### Interface
+
+The interface to `SpectrumInfo` and `DetectorInfo` would be almost identical.
```

Comment:
Not sure what is best design-wise. I would like to avoid having virtual-function resolution.

Maybe more importantly: We do not want clients to throw `DetectorInfo` and `SpectrumInfo` in the same bucket. According to my current understanding, they are two different things and *it should not be possible to threat them in the same way via a base class interface*.